### ISSUE TITLE
Rule Request: Prefer `static` over `final class`

### ISSUE THREAD

**Issue by SimplyDanny** [realm/SwiftLint#5471]
The modifiers `final` and `class` on a method/property have the same meaning as `static`. The rule should flag appearances of `final class` on the same declaration and replace them by `static`.

I don't think there is any need for configuration. The rule could be enabled by default, but this can be decided in review when browsing through the reported OSS findings.

Triggering:

```swift
class C {
    final class func f() {}
}

final class C {
    class func f() {}
}
```

Non-triggering:

```swift
class C {
    static func f() {}
}

class C {
    class func f() {}
}

final class C {}
```

**Comment by omelics** [comment]
The rule conflicts with nested classes. For example:

```swift
extension MyView {
    final class SelectionView: UIView { }
}
```

This obviously cannot be replaced with the `static` keyword and will lead to a compilation error.

**Comment by SimplyDanny** [comment]
> The rule conflicts with nested classes. For example:
> 
> ```swift
> extension A {
>     final class B { }
> }
> ```
> 
> This obviously cannot be replaced with the `static` keyword and will lead to a compilation error.

Which part is replaced by the rule in this example?

**Comment by omelics** [comment]
> > The rule conflicts with nested classes. For example:
> > ```swift
> > extension A {
> >     final class B { }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This obviously cannot be replaced with the `static` keyword and will lead to a compilation error.
> 
> Which part is replaced by the rule in this example?

Sorry, the previous example was incorrect.

The following code is triggered by the rule, while it should not, I believe:

```swift
final class B: UIView {
    override class var layerClass: AnyClass {
        return CALayer.self
    }
}
```

Seems like it happens because of combination of a `final class` and a `class` property override. Swift do allow replacing `class` keyword with `static` in this scenario, but Xcode suggests using `override class var` by default.

**Comment by SimplyDanny** [comment]
> Seems like it happens because of combination of a `final class` and a `class` property override. Swift do allow replacing `class` keyword with `static` in this scenario, but Xcode suggests using `override class var` by default.

Just checked with

```swift
class B {
    class var a: Int { 1 }
}

final class A: B {
    override static var a: Int { 2 }
}
```

which compiles fine in Swift 5.10.

**Comment by omelics** [comment]
> > Seems like it happens because of combination of a `final class` and a `class` property override. Swift do allow replacing `class` keyword with `static` in this scenario, but Xcode suggests using `override class var` by default.
> 
> Just checked with
> 
> ```swift
> class B {
>     class var a: Int { 1 }
> }
> 
> final class A: B {
>     override static var a: Int { 2 }
> }
> ```
> 
> which compiles fine in Swift 5.10.

This compiles fine, you are right. But the following does not.

```swift
final class A: B {
    override class var a: Int { 2 }
}
```

This is why I highlighted the default Xcode suggestion behavior in this case.

And, by the way, should the rule even been triggered by this kind of code? It has no sequential combination of `final class` keywords. It does have a separate `final class` and a separate `class` / `static` property (not `class` or `func`) instead, isn't it?

**Comment by SimplyDanny** [comment]
This is compiles fine, you are right. But the following does not.
> 
> ```swift
> final class A: B {
>     override class var a: Int { 2 }
> }
> ```
> 
> This is why I highlighted the default Xcode suggestion behavior in this case.

What is the error in this example? Both variants seem to be fine for me.

> And, by the way, should the rule even been triggered by this kind of code? It has no sequential combination of `final class` keywords inside the class. It does have a separate `final class` and a separate `class` / `static` property (not `class` or `func`) instead, isn't it?

This is exactly the idea of this rule. It mimics the behavior of the compiler when you use `open` in a `final` class. Using `class` instead of `static` in a `final` class is the same "violation" but for class properties or functions. The purpose of this rule is to serve the same warning in the "static" case.

**Comment by omelics** [comment]
> This is exactly the idea of this rule. It mimics the behavior of the compiler when you use `open` in a `final` class. Using `class` instead of `static` in a `final` class is the same "violation" but for class properties or functions. The purpose of this rule is to serve the same warning in the "static" case.

So let me clarify a little.

Let's say there is the same base class `B`.

```swift
class B {
    class var a: Int { 1 }
}
```

1. The following code will **FAIL** to pass the rule, because the `final class` `A` should not contain a `class` property, and it should be replaced with a `static` one:

```swift 
final class A: B {
    override class var a: Int { 2 }
}
```

2. Thus the following code will **PASS** the rule as expected:

```swift 
final class A: B {
    override static var a: Int { 2 }
}
```

3. Both of the following blocks of code will **PASS** the rule, because the class `A` is not marked as `final`, so it can contain any property you want, `class` or `static`:

```swift
class A: B {
    override class var a: Int { 2 }
}
```
```swift
class A: B {
    override static var a: Int { 2 }
}
```

Is this the intended behavior of the rule?

**Comment by SimplyDanny** [comment]
> Is this the intended behavior of the rule?

Yes, once a class comes with `final` you should not state that something can be overridden by using `class`. Without the `final` it's your decision to put `static` or `class`.

**Comment by omelics** [comment]
Ok, thanks!
